aapcs64-morello/aapcs64-morello.rst

```diff
@@ -810,78 +816,9 @@ The algorithm to implement the generic ``va_arg(ap,type)`` macro is then most ea
-            if (!containsCapabilities(type)) {
-              if (alignof(type) > 8)
-                  offs = (offs + 15) & -16;   // round up
-              nreg = (sizeof(type) + 7) / 8;
-              ap.__gr_offs = offs + (nreg * 8);
-            } else {
-              offs = (offs + 7) & -8;// round up
-              nreg = sizeof(type) / sizeof(void * __capability);
-              ap.__gr_offs = offs + (nreg * 8);
-            }
-            if (ap.__gr_offs > 0)
-                goto on_stack;              // overflowed reg save area
-    #ifdef BIG_ENDIAN
-            if (classof(type) != "aggregate" && sizeof(type) < 8)
-                offs += 8 - sizeof(type);
-    #endif
-            if (containsCapabilities(type)) {
-              // Types containing capabilities are passed in capability
-              // registers. Capability registers are stored in reverse
-              // order.
-              type T;
-              for (unsigned reg = 0; reg < nreg; ++reg) {
-                index = (reg + 1) * 16;
-                ((void *__capability *)&T)[reg] =
-                    *(void *__capability)(ap.__gr_top - 2 * offs - index;
-              }
-              return T;
-            }
-            return *(type *)(ap.__gr_top + offs);
-        } else if (type is an HFA or an HVA) {
-            type ha;       // treat as "struct {ftype field[n];}"
-            offs = ap.__vr_offs;
-            if (offs >= 0)
-                goto on_stack;              // reg save area empty
-            nreg = sizeof(type) / sizeof(ftype);
-            ap.__vr_offs = offs + (nreg * 16);
-            if (ap.__vr_offs > 0)
-                goto on_stack;              // overflowed reg save area
-    #ifdef BIG_ENDIAN
-            if (sizeof(ftype) < 16)
-                offs += 16 - sizeof(ftype);
-    #endif
-            for (i = 0; i < nreg; i++, offs += 16)
-                ha.field[i] = *((ftype *)(ap.__vr_top + offs));
-            return ha;
-        } else if (type passed in fp/simd registers) {
-            offs = ap.__vr_offs;
-            if (offs >= 0)
-                goto on_stack;              // reg save area empty
-            nreg = (sizeof(type) + 15) / 16;
-            ap.__vr_offs = offs + (nreg * 16);
-            if (ap.__vr_offs > 0)
-                goto on_stack;              // overflowed reg save area
-    #ifdef BIG_ENDIAN
-            if (classof(type) != "aggregate" && sizeof(type) < 16)
-                offs += 16 - sizeof(type);
-    #endif
-            return *(type *)(ap.__vr_top + offs);
-        }
-    on_stack:
-        intptr_t arg = ap.__stack;
-        if (alignof(type) > 8)
-            arg = (arg + 15) & -16;
-        ap.__stack = (void *)((arg + sizeof(type) + 7) & -8);
-    #ifdef BIG_ENDIAN
-        if (classof(type) != "aggregate" && sizeof(type) < 8)
-            arg += 8 - sizeof(type);
-    #endif
+        intptr_t arg = (intptr_t)ap.__stack;
+        if (alignof(type) > 16 || sizeof(type) > 16)
```

Comment:
Hm. In theory you could pass larger values inline and it's still memory-safe, but I don't know how common it is and thus if the indirection really matters, plus this does provide slightly more type "safety"

Reply:
The reasoning was that if the caller reads the wrong type once, reading subsequent argument with the correct type would still give you the correct value back (since we're only advancing by 16 bytes at a time). Plus, we can get the number of passed arguments this way.

Reply:
The number of passed arguments is only relevant for the Linux syscall ABI, which is designed to avoid indirect arguments anyway, so I don't think that's ever relevant?